File: lib/go/internal/schema/DocumentLocators.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package schema

import (
	"reflect"
	"sync"
)

type DocumentLocator func(path []any) map[string]any

var documentLocatorRegistry sync.Map

func mapIdentity(documentNamesToJSON map[string]string) uintptr {
	return reflect.ValueOf(documentNamesToJSON).Pointer()
}

func SetDocumentLocators(documentNamesToJSON map[string]string, documentLocators map[string]DocumentLocator) {
	if documentNamesToJSON == nil {
		return
	}

	cloned := make(map[string]DocumentLocator, len(documentLocators))
	for key, value := range documentLocators {
		cloned[key] = value
	}
	documentLocatorRegistry.Store(mapIdentity(documentNamesToJSON), cloned)
}

func CopyDocumentLocators(src map[string]string, dst map[string]string) {
	if src == nil || dst == nil {
		return
	}

	value, ok := documentLocatorRegistry.Load(mapIdentity(src))
	if !ok {
		return
	}

	locators, ok := value.(map[string]DocumentLocator)
	if !ok {
		return
	}

	SetDocumentLocators(dst, locators)
}

func ResolveDocumentCoordinates(path []any, documentName string, documentNamesToJSON map[string]string) map[string]any {
	if value, ok := documentLocatorRegistry.Load(mapIdentity(documentNamesToJSON)); ok {
		if locators, ok := value.(map[string]DocumentLocator); ok {
			if locator, ok := locators[documentName]; ok && locator != nil {
				return locator(append([]any{}, path...))
			}
		}
	}

	return GetPathDocumentCoordinatesPseudoJSON(append([]any{}, path...), documentNamesToJSON[documentName])
}

File: lib/ts/src/internal/schema/DocumentLocators.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { getPathDocumentCoordinatesPseudoJson } from './GetPathDocumentCoordinatesPseudoJson.js';

export type Path = (string | number)[];
export type Coordinates = { row: number; col: number };
export type DocumentLocator = (path: Path) => Coordinates;

const DOCUMENT_LOCATORS = Symbol('telepactDocumentLocators');

type DocumentMapWithLocators = Record<string, string> & {
    [DOCUMENT_LOCATORS]?: Record<string, DocumentLocator>;
};

export function setDocumentLocators(
    documentNamesToJson: Record<string, string>,
    documentLocators: Record<string, DocumentLocator>,
): void {
    Object.defineProperty(documentNamesToJson, DOCUMENT_LOCATORS, {
        configurable: true,
        enumerable: false,
        value: { ...documentLocators },
        writable: true,
    });
}

export function getDocumentLocators(
    documentNamesToJson: Record<string, string>,
): Record<string, DocumentLocator> | undefined {
    return (documentNamesToJson as DocumentMapWithLocators)[DOCUMENT_LOCATORS];
}

export function copyDocumentLocators(
    src: Record<string, string>,
    dst: Record<string, string>,
): void {
    const srcLocators = getDocumentLocators(src);
    if (srcLocators !== undefined) {
        setDocumentLocators(dst, srcLocators);
    }
}

export function resolveDocumentCoordinates(
    path: Path,
    documentName: string,
    documentNamesToJson: Record<string, string>,
): Coordinates {
    const documentLocators = getDocumentLocators(documentNamesToJson);
    const locator = documentLocators?.[documentName];
    if (locator !== undefined) {
        return locator(path);
    }

    const document = documentNamesToJson[documentName] ?? '[]';
    return getPathDocumentCoordinatesPseudoJson(path, document);
}
